linux/src/async_socket_linux.c

```diff
@@ -0,0 +1,695 @@
+        for(int index = 0; index < num_ready; index++)
+        {
+            if (events[index].events & EPOLLRDHUP)
+            {
+                ASYNC_SOCKET_RECV_CONTEXT* recv_context = events[index].data.ptr;
+                if (recv_context != NULL)
+                {
+                    recv_context->on_receive_complete(recv_context->on_receive_complete_context, ASYNC_SOCKET_RECEIVE_ABANDONED, 0);
+                }
+            }
+            else if (events[index].events & EPOLLIN)
+            {
+                // Receive data from the socket
+                ASYNC_SOCKET_RECEIVE_RESULT receive_result;
+                ASYNC_SOCKET_RECV_CONTEXT* recv_context = events[index].data.ptr;
+
+                if (recv_context != NULL)
+                {
+                    uint32_t index = 0;
+                    ssize_t recv_size;
+                    do
+                    {
+
+                        recv_size = recv(recv_context->socket_handle, recv_context->recv_buffers[index].buffer, recv_context->recv_buffers[index].length, 0);
+                        if (recv_size < 0)
+                        {
+                            if (errno == EAGAIN || errno == EWOULDBLOCK)
+                            {
+                                // No more data to recv
+                                receive_result = ASYNC_SOCKET_RECEIVE_OK;
+                                break;
+                            }
+                            else
+                            {
+                                if (errno == ECONNRESET)
+                                {
+                                    receive_result = ASYNC_SOCKET_RECEIVE_ABANDONED;
+                                }
+                                else
+                                {
+                                    // Error here
+                                    receive_result = ASYNC_SOCKET_RECEIVE_ERROR;
+                                    LogError("failure recv error: %d", errno);
+                                }
+                                break;
+                            }
+                        }
+                        else if (recv_size == 0)
+                        {
+                            receive_result = ASYNC_SOCKET_RECEIVE_OK;
+                            break;
+                        }
+                        else
+                        {
+                            if (index + 1 >= recv_context->total_buffer_count || recv_size < recv_context->recv_buffers[index].length)
+                            {
+                                // We gotten all the data we need to get or we don't have any more recv space
+                                receive_result = ASYNC_SOCKET_RECEIVE_OK;
+                                break;
+                            }
+                            else
+                            {
+                                index++;
```

Comment:
This is interesting.
I think we should actually indicate the data we got, right?
We're sort of trying to fill more buffers and now the data we got is lingering in our stack instead of being processed.

Reply:
This is interesting because the async_socket_receive_async takes in an array of ASYNC_SOCKET_BUFFER.  So you can only receive half the size at a time.  I'm guessing we could call the callback after each recv, but then how to tell the callback to look at the 2nd array position?